src/systems/a320_systems/src/flight_warning/ecam_control.rs

```diff
@@ -0,0 +1,166 @@
+
+    clear_1: EcamControlPanelButton,
+    status: EcamControlPanelButton,
+    recall: EcamControlPanelButton,
+    clear_2: EcamControlPanelButton,
+    to_config_test: EcamControlPanelButton,
+    emergency_cancel: EcamControlPanelButton,
+
+    warning_switches_word: Arinc429Word<u32>,
+}
+
+impl A320EcamControlPanel {
+    const ARINC_429_REFRESH_INTERVAL: Duration = Duration::from_millis(60);
+
+    pub fn new(context: &mut InitContext, powered_by: ElectricalBusType) -> Self {
+        Self {
+            powered_by,
+            is_powered: context.has_engines_running(),
+            stale_for: Self::ARINC_429_REFRESH_INTERVAL,
+            clear_1_id: context.get_identifier("BTN_CLR".to_owned()),
+            status_id: context.get_identifier("BTN_STS".to_owned()),
+            recall_id: context.get_identifier("BTN_RCL".to_owned()),
+            clear_2_id: context.get_identifier("BTN_CLR2".to_owned()),
+            to_config_test_id: context.get_identifier("BTN_TOCONFIG".to_owned()),
+            emergency_cancel_id: context.get_identifier("BTN_EMERCANC".to_owned()),
+            clear_1: EcamControlPanelButton::default(),
+            status: EcamControlPanelButton::default(),
+            recall: EcamControlPanelButton::default(),
+            clear_2: EcamControlPanelButton::default(),
+            to_config_test: EcamControlPanelButton::default(),
+            emergency_cancel: EcamControlPanelButton::default(),
+            warning_switches_word: Arinc429Word::new(0, SignStatus::FailureWarning),
+        }
+    }
+
+    pub fn update(&mut self, context: &UpdateContext) {
+        if !self.is_powered {
+            self.warning_switches_word = Arinc429Word::new(0, SignStatus::FailureWarning);
+            self.stale_for = Self::ARINC_429_REFRESH_INTERVAL;
+            self.clear_1.update(false);
+            self.status.update(false);
+            self.recall.update(false);
+            self.clear_2.update(false);
+            self.to_config_test.update(false);
+            self.emergency_cancel.update(false);
+            return;
+        }
+
+        if self.stale_for >= Self::ARINC_429_REFRESH_INTERVAL {
+            let mut word = Arinc429Word::new(0, SignStatus::NormalOperation);
+            word.set_bit(11, self.clear_1.read());
+            word.set_bit(13, self.status.read());
+            word.set_bit(14, self.recall.read());
+            word.set_bit(16, self.clear_2.read());
+            word.set_bit(18, self.to_config_test.read());
+            self.warning_switches_word = word;
+
+            self.clear_1.update(true);
+            self.status.update(true);
+            self.recall.update(true);
+            self.clear_2.update(true);
+            self.to_config_test.update(true);
+            self.emergency_cancel.update(true);
+            while self.stale_for >= Self::ARINC_429_REFRESH_INTERVAL {
+                self.stale_for -= Self::ARINC_429_REFRESH_INTERVAL;
+            }
```

Comment:
It's to ensure the overall timings stay in-tact - I can add a comment.

As an example: Basically imagine we update again and again with deltaTime 21. Now after 12 ticks have elapsed were at an accumulated delta time of 252 ms. If we reset it to 0, it's going to take another 252 ms for the next full update. Long-term that means our refresh interval is too infrequent.

With this code we effectively run elapsed time modulo 240ms. If you repeat the example above we're still unlucky and the first system tick takes 252 ms. But afterwards we start with an accumulated 12 ms to compensate, and after 11 ticks we're at 243ms - much better! This minimizes the long-term error.

Even better: carrying the remainder forward allows us to model very slight time drift. You can see above that I subtract a few micro seconds from the refresh time so it's not exactly 240s. This is so that over a flight over multiple hours the FWCs drift very slightly to each other, allowing timing differences like the separate MW/MC illumination to not always be exactly the same - it's like two cars with blinkers at the same time, or two clocks next to each other. Short of a shared time source one will invariably drift by a few milliseconds from the other.

I will add comments to explain this.

Reply:
Great explanation, thanks! I like this concept.